File set combinators base: `lib.fileset.toSource` (#245623)
NixOS/nixpkgs: lib/fileset/README.md

    What should the behavior be if `./foo` itself is excluded but all of its contents are included?
    It leads to having to define when directories are recursed into, but then we're effectively back at how the `builtins.path`-based filters work.

  - `./.` represents all files in `./.` _and_ the directory itself, but not its subdirectories, meaning that at least `./.` will be preserved even if it's empty.

    In that case, `intersect ./. ./foo` should only include files and no directories themselves, since `./.` includes only `./.` as a directory, and same for `./foo`, so there's no overlap in directories.
    But intuitively this operation should result in the same as `./foo`, that not being the case is just confusing.
- (+) This matches Git also only supporting files, so developers are already used to it
- (-) Empty directories (even if they contain nested directories) are neither representable nor preserved when coercing from paths.
  - (+) It is very rare that empty directories are necessary.
  - (+) We can implement a workaround, allowing `toSource` to take an extra argument for ensuring certain extra directories exist in the result.

> - (-) It slows down store imports, since the evaluator needs to traverse the entire tree to remove any empty directories

Author: Having done some brief benchmarks, it turns out this might actually be very significant. While it does make the design nice to strictly never have empty directories, in practice almost nobody uses or cares about empty directories. So we'd be paying a significant performance cost for an insignificant benefit. It could be optimized in Nix itself, but even that takes considerable effort with the same insignificant benefit.

The one decent alternative I can think of is to not remove empty directories within directories that don't have to be recursed into. So e.g. `./.` as a fileset would not remove any empty directories. but `fileFilter pred ./.` would, because `fileFilter` has to recurse into the entire thing.

I will add a proper benchmark for this so we can reason about this easier.

Author: Thinking about this some more, I don't think that alternative is okay. It leads to unexpected and hard-to-explain semantics which we could not get away from. The right solution is to have better support for this in Nix itself, e.g. by adjusting `builtins.path` as I just suggested here: https://github.com/NixOS/nix/issues/8820

Author: Note to self: Reuse the benchmark and Nix code from https://github.com/NixOS/nix/issues/8820#issuecomment-1678256565, maybe even export the Nix code as a new function under `lib.sources`, since it's useful on its own and could use a bunch of testing, the implementation is very non-trivial. It would also remove the nastiest code from the file set combinators implementation.

Author: Alright this should be fine, as discussed in the above issue, it's I/O bottlenecked, and while there is increased memory usage, it's not _too_ bad. I also added a benchmark now, allowing comparing to previous versions.

- (+) It allows adjustments to the representation, which is especially useful in the early days of the library.
- (+) It still allows the representation to be stabilized later if necessary and it has proven itself

### Influence tracking

File set operations internally track the top-most directory that could influence the included files.
And `toSource` requires `fileset` to not be influencable by files outside of `root`.
For example, even with `dir/file.txt` being the only file in `./.`, `toSource { root = ./dir; fileset = ./.; }` gives an error.
This is because the file set could be influenced by any files in `./.`, which is outside `./dir`.

Arguments:
- (+) This gives us the guarantee that adding new files to a project never breaks a file set expression.
  This also exists in a lesser form for removed files, where only removing files referenced by paths coerced to file sets can break a file set expression.
- (+) This can be removed later, if we discover it's too restrictive
- (-) It leads to errors when a sensible result could be returned, such as in the above example.

### Empty directories

> File sets strictly can only represent a _set_ of local files, directories on their own are not representable.

Reviewer: ```suggestion
File sets can only represent a _set_ of local files, directories on their own are not representable.
```

  Here's some possibilities:
  - `./.` represents the files in `./.` _and_ the directory itself including its subdirectories, meaning that even if there's no files, the entire structure of `./.` is preserved

    In that case, what should `fileFilter (file: false) ./.` return?
    It could return the entire directory structure unchanged, but with all files removed, which would not be what one would expect.

    Trying to have a filter function that also supports directories will lead to the question of:
    What should the behavior be if `./foo` itself is excluded but all of its contents are included?
    It leads to having to define when directories are recursed into, but then we're effectively back at how the `builtins.path`-based filters work.

  - `./.` represents all files in `./.` _and_ the directory itself, but not its subdirectories, meaning that at least `./.` will be preserved even if it's empty.

    In that case, `intersect ./. ./foo` should only include files and no directories themselves, since `./.` includes only `./.` as a directory, and same for `./foo`, so there's no overlap in directories.

>     But intuitively this operation should result in the same as `./foo`, that not being the case is just confusing.

Reviewer: ```suggestion
    But intuitively this operation should result in the same as `./foo` – everything else is just confusing.
```

> - (+) This matches Git also only supporting files, so developers are already used to it

Reviewer: ```suggestion
- (+) This matches how Git only supports files, so developers should already be used to it.
```

- (-) Empty directories (even if they contain nested directories) are neither representable nor preserved when coercing from paths.
  - (+) It is very rare that empty directories are necessary.
  - (+) We can implement a workaround, allowing `toSource` to take an extra argument for ensuring certain extra directories exist in the result.
- (-) It slows down store imports, since the evaluator needs to traverse the entire tree to remove any empty directories
  - (+) This can still be optimized by introducing more Nix builtins if necessary

### String paths

File sets do not support Nix store paths in strings such as `"/nix/store/...-source"`.

Arguments:
- (+) Such paths are usually produced by derivations, which means `toSource` would either:

>   - Require IFD if `builtins.path` is used as the underlying primitive

Reviewer: ```suggestion
  - Require [Import From Derivation](../../pkgs/README.md#import-from-derivation) if `builtins.path` is used as the underlying primitive
```

Author: Hmm that section is only saying that IFD is disallowed in Nixpkgs, but the file set library isn't really meant to be used in Nixpkgs. I'd say let's leave it at "IFD" for now and wait for https://github.com/NixOS/nix/pull/7332

  - Require importing the entire `root` into the store such that derivations can be used to do the filtering
- (+) The convenient path coercion like `union ./foo ./bar` wouldn't work for absolute paths, requiring more verbose alternate interfaces:

>   - `let root = "/nix/store/...-source"; in union "${root}/foo" "${root}/bar"`:
>     Verbose and dangerous because if `root` was a path, the entire path would get imported into the store.
>   - `toSource { root = "/nix/store/...-source"; fileset = union "./foo" "./bar"; }`:
>     Does not allow debug printing intermediate file set contents, since we don't know the paths contents before having a `root`.
>   - `let fs = lib.fileset.withRoot "/nix/store/...-source"; in fs.union "./foo" "./bar"`:
>     Makes library functions impure since they depend on the contextual root path, questionable composability.
> - (+) The point of the file set abstraction is to specify which files should get imported into the store.

Reviewer: ```suggestion
  - `let root = "/nix/store/...-source"; in union "${root}/foo" "${root}/bar"`

    Verbose and dangerous because if `root` was a path, the entire path would get imported into the store.

  - `toSource { root = "/nix/store/...-source"; fileset = union "./foo" "./bar"; }`

    Does not allow debug printing intermediate file set contents, since we don't know the paths contents before having a `root`.

  - `let fs = lib.fileset.withRoot "/nix/store/...-source"; in fs.union "./foo" "./bar"`
 
    Makes library functions impure since they depend on the contextual root path, questionable composability.

- (+) The point of the file set abstraction is to specify which files should get imported into the store.

```

That should be a bit nicer to read.

# File set design

## File set data structure

```
fileset :: {
  # Tag to indicate this value is a file set
  _type = "fileset";

> 

Author: Now added in https://github.com/NixOS/nixpkgs/pull/245623/commits/21afdcd38c5aed316a9c876abe6758d77688723c

  # Only files under this path can be represented
  # Always a directory
  _internalBase :: path;

  # A tree representation of all included files
  _internalTree :: filesetTreeDir;
};
```

## `filesetTree`: A directory entry value

> 

Author: I'm now using more natural language to describe it without custom syntax: https://github.com/tweag/nixpkgs/blob/fileset.toSource/lib/fileset/README.md#internal-representation
